lib/rubocop/cop/mixin/hash_slice_except.rb

```diff
@@ -0,0 +1,160 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    # Common functionality for Style/HashExcept and Style/HashSlice
+    module HashSliceExcept
+      include RangeHelp
+      extend NodePattern::Macros
+
+      RESTRICT_ON_SEND = %i[reject select filter].freeze
+
+      # @!method bad_method_with_poro?(node)
+      def_node_matcher :bad_method_with_poro?, <<~PATTERN
+        (block
+          (call _ _)
+          (args
+            $(arg _)
+            (arg _))
+          {
+            $(send
+              _ {:== :!= :eql? :include?} _)
+            (send
+              $(send
+                _ {:== :!= :eql? :include?} _) :!)
+            })
+      PATTERN
+
+      # @!method bad_method_with_active_support?(node)
+      def_node_matcher :bad_method_with_active_support?, <<~PATTERN
+        (block
+          (send _ _)
+          (args
+            $(arg _)
+            (arg _))
+          {
+            $(send
+              _ {:== :!= :eql? :in? :include? :exclude?} _)
+            (send
+              $(send
+                _ {:== :!= :eql? :in? :include? :exclude?} _) :!)
+            })
+      PATTERN
+
+      def on_send(node)
+        raise NotImplementedError
+      end
+
+      private
+
+      def extract_offense(node)
+        block = node.parent
+        return unless bad_method?(block)
+
+        except_key = except_key(block)
+        return if except_key.nil? || !safe_to_register_offense?(block, except_key)
+
+        [offense_range(node), except_key_source(except_key)]
+      end
+
+      # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+      def bad_method?(block)
+        if active_support_extensions_enabled?
+          bad_method_with_active_support?(block) do |key_arg, send_node|
+            return false if send_node.method?(:in?) && send_node.receiver&.source != key_arg.source
+            return true if !send_node.method?(:include?) && !send_node.method?(:exclude?)
+
+            send_node.first_argument&.source == key_arg.source
+          end
+        else
+          bad_method_with_poro?(block) do |key_arg, send_node|
+            !send_node.method?(:include?) || send_node.first_argument&.source == key_arg.source
+          end
+        end
+      end
+      # rubocop:enable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+      def semantically_except_method?(send)
+        block = send.parent
+        body = block.body
+
+        negated = body.method?('!')
+        body = body.receiver if negated
+
+        case send.method_name
+        when :reject
+          body.method?('==') || body.method?('eql?') || included?(negated, body)
+        when :select, :filter
+          body.method?('!=') || not_included?(negated, body)
+        end
+      end
+
+      def semantically_slice_method?(send)
+        semantically_except_method?(send) == false
+      end
```

Comment:
Is there any reason `semantically_except_method?(send)` would return nil? Can this just be `!semantically_except_method?(send)`?

In other words, at the point where this is called, can we guarantee that the code should either become a `slice` or `except` call?